### PR TITLE
[PD] improve blind hole depth handling

### PR DESCRIPTION
**Pull request by donovaly**
as reported in https://tracker.freecadweb.org/view.php?id=3818
we treat the blind hole depth not according to the conventions. The size of the drill point due to the angle is normally not taken into account but FC does this in every case.

There are of course cases where the total depth, including the drill point size matters.

This PR adds therefore an option, that is by default off, to take the drill point size into account.
Without the option, (the new default) the depth is nor calculated according to the conventions.

I purposely made the new option by default off, because all literature about technical drawings don't take the drill point size into account and the hole dialog should by default offer the normed / conventioned behavior.

As side effect that means that existing blind holes change, but the same applied for my last hole dialog fixed (diameters of existing holes changed). So I think it is OK. If you disagree, please tell. However, I don't see how this can be avoided.

Forum thread: https://forum.freecadweb.org/viewtopic.php?t=51491&view=unread#p457068

### CODE REVIEW

**Comment by chennes** [comment]
When I tried to test this I got a crash due to a debug assertion failure in OCC. The steps I followed were:
1. Create a rectangle in a sketch
2. Pad the rectangle 50mm
3. Sketch a circle on that pad's upper surface
4. Activate the hole task
5. Check the new option to "Take into account for depth" (without changing anything else).

Error message is:
"Error: a duplicated modification of a shape. (Condition: "!aModitications->Contains(theModified)")"

Backtrace:
```
 	TKBRepd.dll!00007ffbbe5fa660()	Unknown
 	TKBOd.dll!00007ffbbf98f2a1()	Unknown
 	TKBOd.dll!00007ffbbf965374()	Unknown
 	TKBOd.dll!00007ffbbf9723b3()	Unknown
 	TKBOd.dll!00007ffbbf970751()	Unknown
 	TKBOd.dll!00007ffbbf910d42()	Unknown
 	TKBOd.dll!00007ffbbf90cc1e()	Unknown
 	TKBOd.dll!00007ffbbf914edc()	Unknown
>	_PartDesign_d.pyd!PartDesign::Hole::execute() Line 1518	C++
 	FreeCADApp_d.dll!App::DocumentObject::recompute() Line 111	C++
 	Part_d.pyd!Part::Feature::recompute() Line 101	C++
 	FreeCADApp_d.dll!App::Document::_recomputeFeature(App::DocumentObject * Feat) Line 3744	C++
 	FreeCADApp_d.dll!App::Document::recomputeFeature(App::DocumentObject * Feat, bool recursive) Line 3804	C++
 	PartDesignGui_d.pyd!PartDesignGui::TaskFeatureParameters::recomputeFeature() Line 74	C++
 	PartDesignGui_d.pyd!PartDesignGui::TaskHoleParameters::drillForDepthChanged() Line 387	C++
 	PartDesignGui_d.pyd!PartDesignGui::TaskHoleParameters::qt_static_metacall(QObject * _o, QMetaObject::Call _c, int _id, void * * _a) Line 176	C++
 	[External Code]	
 	FreeCADGui_d.dll!Gui::GUIApplication::notify(QObject * receiver, QEvent * event) Line 92	C++
 	[External Code]	
 	FreeCADGui_d.dll!Gui::GUIApplication::notify(QObject * receiver, QEvent * event) Line 92	C++
 	[External Code]	
 	FreeCADGui_d.dll!Gui::Application::runApplication() Line 2278	C++
 	FreeCAD_d.exe!main(int argc, char * * argv) Line 302	C++
 	FreeCAD_d.exe!WinMain(HINSTANCE__ * __formal, HINSTANCE__ * __formal, char * __formal, int __formal) Line 97	C++
 	[External Code]	

```

**Comment by wwmayer** [comment]
I cannot confirm the behaviour you observed but I realized that there are several locations where the doubles x and z are declared and used for further calculations but they are never correctly initialized so behaviour is undefined.

These are the lines 1269, 1298 and 1349.

File: src/Mod/PartDesign/App/FeatureHole.cpp [hunk 1]
```diff
@@ -1366,8 +1397,8 @@ App::DocumentObjectExecReturn *Hole::execute(void)
             mkWire.Add(BRepBuilderAPI_MakeEdge(lastPoint, newPoint));
             lastPoint = newPoint;
```
Comment (chennes) [inline src/Mod/PartDesign/App/FeatureHole.cpp:1398]: Is it correct to have these lines here? They look like the tip formation code that's now in the conditional above.

File: src/Mod/PartDesign/App/FeatureHole.cpp [hunk 2]
```diff
@@ -1346,14 +1348,43 @@ App::DocumentObjectExecReturn *Hole::execute(void)
...
+                mkWire.Add(BRepBuilderAPI_MakeEdge(lastPoint, newPoint));
+                lastPoint = newPoint;
+
+                // the end point is the size of the drill tip
+                newPoint = toPnt((-length - radius * tan(drillPointAngle)) * zDir);
+                mkWire.Add(BRepBuilderAPI_MakeEdge(lastPoint, newPoint));
+                lastPoint = newPoint;
+            }
 
             if (z > 0 || z >= length1)
```
Comment (chennes) [inline src/Mod/PartDesign/App/FeatureHole.cpp:1389]: `length1` is not a very descriptive variable name -- what is it?

File: src/Mod/PartDesign/App/FeatureHole.cpp [hunk 3]
```diff
@@ -1346,14 +1348,43 @@ App::DocumentObjectExecReturn *Hole::execute(void)
...
-                                gp_Pnt(radius, 0, 0),
-                                gp_Pnt(radiusBottom, -length, 0), x, z);
+            bool isDrillForDepth = DrillForDepth.getValue();
+
+            // the angle is in any case > 0 and < 90 but nevertheless this safeguard:
+            if ( drillPointAngle <= 0 || drillPointAngle >= Base::toRadians( 180.0 ) )
+                return new App::DocumentObjectExecReturn("Hole error: Invalid drill point angle");
+
+            // if option to take drill point size into account
+            // the next wire point is the intersection of the drill dege and the hole edge
```
Comment (chennes) [inline src/Mod/PartDesign/App/FeatureHole.cpp:1358]: Typo - dege -> edge

Comment (donovaly) [reply src/Mod/PartDesign/App/FeatureHole.cpp]: fixed

File: src/Mod/PartDesign/App/FeatureHole.cpp (hunk 2 above)
Comment (donovaly) [reply src/Mod/PartDesign/App/FeatureHole.cpp]: Yes, this code is hard to read because of these meaningless variables. However, it is not my code. 
Nevertheless I renamed this variable to "lengthCounter" to make clear that it is the length of the counter(sink/bore).

File: src/Mod/PartDesign/App/FeatureHole.cpp (hunk 1 above)
Comment (donovaly) [reply src/Mod/PartDesign/App/FeatureHole.cpp]: Thanks, this was a leftover from copy paste (doubled code). It should not be there. Fixed now.

**Comment by donovaly** [comment]
> there are several locations where the doubles x and z are declared and used for further calculations but they are never correctly initialized so behaviour is undefined.

Yes, and since their meaningless name it is even hard to follow what the code does. I now renamed the variables and initialized them.

I also improved the code a little bit here. But the whole hole code could benefit from some love. However, not for now.

**Comment by donovaly** [comment]
> When I tried to test this I got a crash due to a debug assertion failure in OCC. The steps I followed were:

I could not reproduce this with a debug build. But maybe it was because of the uninitialized variables. Can you please check again?

**Comment by chennes** [comment]
> 
> 
> > When I tried to test this I got a crash due to a debug assertion failure in OCC. The steps I followed were:
> 
> I could not reproduce this with a debug build. But maybe it was because of the uninitialized variables. Can you please check again?

It seems that it was being caused by the weird geometry that was generated when the tip doubled back on itself (due to those duplicated lines 1389 through 1398) -- a3bcf88 should do the trick, but I will re-test shortly.

**Comment by donovaly** [comment]
Can this PR go in? I fixed the 2 issues spotted by @chennes . I see that the PR is necessary since it popped up again in the forum: https://forum.freecadweb.org/viewtopic.php?t=51491&view=unread#p462138

**Review by chennes — APPROVED** [review]
I just re-tested and the bug I reported is resolved. It works well and is a good addition, thanks!

**Comment by donovaly** [comment]
> Can this PR go in?

Ping.

@chennes tested it and we have another PR (#4274) that waits for it. (Waits for this PR because it will change the Hole dialog further.)

File: src/Mod/PartDesign/Gui/TaskHoleParameters.ui [hunk 4]
```diff
@@ -20,28 +20,28 @@
...
-   <item row="0" column="0" colspan="2">
-    <widget class="QLabel" name="label_13">
-     <property name="text">
-      <string>&lt;b&gt;Threading and size&lt;/b&gt;</string>
-     </property>
-    </widget>
-   </item>
-   <item row="1" column="0">
-    <widget class="QLabel" name="label_2">
+   <item row="7" column="3" colspan="2">
```
Comment (davidosterberg) [inline src/Mod/PartDesign/Gui/TaskHoleParameters.ui:23]: Please sort this ui file ;)

**Review by davidosterberg — CHANGES_REQUESTED** [review]
I have not tested it, but the code looks good. I request that the UI file is sorted, and the PR rebased. Otherwise this looks ready (considering @chennes approval).

**Comment by donovaly** [comment]
>  I request that the UI file is sorted, and the PR rebased. Otherwise this looks ready (considering @chennes approval).

Done, thanks for having a look.

**Review by davidosterberg — APPROVED** [review]
Looks good to me

**Comment by donovaly** [comment]
Many thanks for merging. We can now proceed to finish the other 2 pending hole PRs.
